Fix Rust caching issues

calavera opened the pull request:
Fixes https://github.com/netlify/build-image/issues/508
Fixes #505

calavera committed:
Cache rust artifacts only when cargo files are detected

This way we don't turn the `target` directory into a reserved directory if people are not using rust.

vbrown608 commented on run-build-functions.sh:
```diff
@@ -685,9 +685,13 @@ cache_artifacts() {
-  cache_home_directory ".cargo/registry" "rust cargo registry cache"
-  cache_home_directory ".cargo/bin" "rust cargo bin cache"
+  
+  if if [ -f Cargo.toml ] || [ -f Cargo.lock ]
```
```suggestion
  if [ -f Cargo.toml ] || [ -f Cargo.lock ]
```

calavera committed:
Update run-build-functions.sh

Co-authored-by: Vivian Brown <vivian@netlify.com>

calavera committed:
Cache the target directory only when a cargo file exists.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Copy target directory using reflinks.

This allows keeping both directories until the deploy is completed,
while keeping the fast nature of moving things around.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
I've opted for changing only the behavior of how we cache the `target` directory, instead of all cached directories to minimize the impact of these changes.

calavera commented:
@vbrown608 I've tested this with an internal build and it seems to work as expected

mraerino commented on run-build-functions.sh:
```diff
@@ -685,9 +689,13 @@ cache_artifacts() {
+  
+  if [ -f Cargo.toml ] || [ -f Cargo.lock ]
+  then
+    cache_home_directory ".rustup" "rust rustup cache"
```
I think we should try to cache this regardless of whether there are Cargo config files

calavera replied:
updated

calavera committed:
Cache .rustup regardless of whether there are cargo files or not.

Signed-off-by: David Calavera <david.calavera@gmail.com>

mraerino reviewed (approved):
awesome!
